### ISSUE TITLE
Why the gradient w.r.t. the parameters before update is used in query set?

### ISSUE THREAD

**Issue by machanic** [katerakelly/pytorch-maml#18]
In this line: https://github.com/katerakelly/pytorch-maml/blob/master/src/inner_loop.py#L76
why don't use `grads = torch.autograd.grad(loss, fast_weights.values())` ? but `self.paramters()` which is the original parameters before any inner-update step.

**Comment by katerakelly** [comment]
The meta-gradients are computed with respect to the pre-update parameters. Intuitively, this is because during meta-training we want to adjust the pre-update parameters such that the post-update parameters are useful for the task.

**Comment by tanqi724** [comment]
Thank your very much for your code. I have a similar concern:

In this line: [https://github.com/katerakelly/pytorch-maml/blob/master/src/inner_loop.py#L76](url)
`# Compute the meta gradient and return it

        in_, target = val_loader.__iter__().next()

        loss,_ = self.forward_pass(in_, target, fast_weights) 

        loss = loss / self.meta_batch_size # normalize loss

        grads = torch.autograd.grad(loss, self.parameters())`

The loss is computed using fast_weights. So why (self.parameters) is used when doing the autograd w.r.t. this loss? That actually causes some bugs when I try to re-implement this function. Am I missing something？
